### PR TITLE
Fix compatibility.h inclusion and add CAML_NAME_SPACE versioning

### PR DESCRIPTION
**Pull request by gadmm**
This was originally meant as a comment to https://github.com/ocaml/ocaml/pull/1798, but in the course of writing it I realised that this fixes an issue with `compatibility.h` for 4.10, so I decided to upgrade it to a pull request. (In case my approach is not found to be consensual, then I intend that it still serves as bug report.)

### Issue

To begin with, the issue I noticed in 4.10: a lot of macros are now included for backwards-compatibility in the file `compatibility.h`. However, the way users automatically get the benefits of `compatibility.h` is quite irregular. Indeed, it is included automatically in most caml headers _when `CAML_NAME_SPACE` is unset_; it is also included automatically in `caml/memory.h` and  `caml/minor_gc.h` _when either `CAML_NAME_SPACE` or `CAML_INTERNALS` is unset_. Thus, users do not automatically get the benefits of `compatibility.h`, unless they forget to define `CAML_NAME_SPACE` (which docs says they should), or if they happen to include `caml/memory.h`. I want to fix this irregularity by letting them automatically benefit from `compatibility.h` on a similar principle as before 4.10. To do so I repurpose the user-defined `CAML_NAME_SPACE` macro, whose initial purpose is to control exactly that. Only, now it gets a version number.

(What about the practice of including `compatibility.h` by hand? Opam audit suggests that it is marginal before 4.10, as the only true positives I have found are `bap` and `imagemagick`.)

### Version macros

For context, see #1798 and the discussion about macros `CAML_API_VERSION` and `CAML_API_COMPATIBILITY`. There is much in common with the patch at #1798. But there seemed to be doubts about the semantics and purpose of the latter. My macro `CAML_NAME_SPACE` is meant to replace it, I will explain the semantics and purpose below, and I think we need this in 4.10.

First, it is not about permitting weaker backwards compatibility requirements (to quote the title of #1798). As a backwards compatibility helper, we have `OCAML_VERSION` (playing the role of `CAML_API_VERSION` in the sense of the latests comments there, i.e. "the actual version number defined by the runtime's headers"). I leave this one aside as we already have it.

Second, about reusing the macro name `CAML_NAME_SPACE`: no strong opinion here, but the following rationale:

1. Not introducing yet another versioning macro;
2. I like it more than the `CAML_API_*` names because it emphasizes that _it is about syntax and not semantics_.

### Meaning of `CAML_NAME_SPACE=X`

Defined by the user, means _“please rid me of these troublesome identifiers that are obsolete in version X”_. It is the user asserting that the program is written for OCaml ≥ X. By default (undefined) it is the lowest possible, i.e. all compatibility helpers defined. It does _not_ mean "please try to respect the API of version X".

(Terminology: obsolete does not mean deprecated. But it can also help with deprecation since it gives a mean to the user to prepare to removal ahead of time.)

### Purpose

It is meant to solve the following problem: imagine you want to track down obsolete identifiers to modernize your code. Before 4.10 you would just compile with "-DCAML_NAME_SPACE", which rids you of `compatibility.h`. Now with the new compatibility helpers, we want to still include this header for people who already define `CAML_NAME_SPACE`, so that the helpers continue to be defined automatically and we can hope that the user has nothing to change to their code. However, as of now there is no mechanism to let the user turn off the new compatibility helpers. The new mechanism lets the user do that by increasing `CAML_NAME_SPACE` to the version of their choice.

Like it was intended with #1798, it can be re-used to fit the purpose of the versioning macro introduced in #1003.

### Notes

* The value `CAML_NAME_SPACE=1`, i.e. with `-DCAML_NAME_SPACE`, of course keeps its current meaning.
* Clean-up redundant inclusions. `compatibility.h` remains included everywhere (via `config.h`).
* The OCaml compiler always enforces the latest name space for itself, via the definition `CAML_NAME_SPACE=99999`.
* Automatically-generated changes to `configure`: I do not know what to do of them.
* Optionally, we can add other conditional definitions, e.g. `Begin/end_roots`, like #1798.

(cc @stedolan for the multicore api version business, pinging also @gasche since you told me you are interested in `compatiblity.h` issues.)

### CODE REVIEW

**Comment by gadmm** [comment]
* Rebased on top of other `compatibility.h` improvements
* ~Removed 40300 value for CAML_NAME_SPACE which cannot be used in practice~
* Fixed the case where the user `#define CAML_NAME_SPACE` without value

**Comment by gadmm** [comment]
Split in two commits to clarify:

* The first one fixes the issue per se and should be in 4.10 I think.
* The seconds one answers the question “How are you going to turn off these definitions (in particular for the compiler itself)?”. Optional for 4.10, but recommended.

**Comment by gadmm** [comment]
About the test failure: this is a bit “interesting”. The option `-DCAML_NAME_SPACE=99999` is passed in Makefiles to compile OCaml, but it is also recorded in `ocamlc` and `ocamlopt` themselves inside `Config.mkexe`. This means that the option is passed by `ocamlc` for linking the custom runtime and by `ocamlopt` for linking. This in turn interferes with user's options they passed via `-ccopt`, e.g. if they do `-DCAML_NAME_SPACE=X` for a different value.

According to commit history, there is a chance that this is unintended given that the option `-DCAML_NAME_SPACE` has been added for the purpose of removing compatibility macros in the compiler itself. I have a hack that fixes that, but if the principle of the PR is accepted then we should look into a proper fix (and look if this hides other similar issues).

**Comment by stedolan** [comment]
I like this proposal. What `CAML_NAME_SPACE` does here is clearer than `CAML_API_VERSION` elsewhere, and reusing the same macro fits.

I have an alternative to `-DCAML_NAME_SPACE=99999`: how about having `compatibility.h` consider `CAML_NAME_SPACE` to be very large whenever `CAML_INTERNALS` is defined. This would fix your testcase issue (the runtime files already define `CAML_INTERNALS`), and I think it would have the right behaviour for user code: if users use the documented public API of OCaml, then they enjoy backwards-compatibility support, but if they poke undocumented internals then they must adapt to runtime changes themselves.

In other words, I think the boundary between "internal code that must be up to date" and "user code that gets compatibility defines" should not be "does this ship with the compiler distribution?" but instead "does this define CAML_INTERNALS?".

**Comment by gadmm** [comment]
The problem is that `-DCAML_NAME_SPACE` is meant to affect all C files including `otherlibs/`, and not all define `CAML_INTERNALS`. This was the explicit purpose of the generalisation of `-DCAML_NAME_SPACE` in #892. 

(edit: reading your last paragraph again, maybe you mean to change that.)

@shindere, do you have an opinion about the `-DCAML_NAME_SPACE=99999` issue described in an earlier comment?

**Comment by gadmm** [comment]
I have included my hack, the idea is to separate the MKEXE used for building the compiler from the one recorded in `ocamlc` and `ocamlopt`. There is probably a nicer way of doing, I am just not familiar enough with the build system.

- [ ] Fix the `-DCAML_NAME_SPACE=9999` issue
- [x] Document

**Comment by gadmm** [comment]
I have included a test. It succeeds with 4.09 and the current PR, and fails with current trunk and 4.10.

The test relies on the fact that `caml_compare_unordered` is public, which is the only one in this situation. For internal runtime state variables, one can write similar tests modulo the definition of `CAML_INTERNALS`. In that case it is legitimate to ask users to fix their code themselves, but the interest in the current PR is that users would fix their code differently depending on whether this PR is accepted (e.g. whether they do not have to include `compatibility.h` by hand).

**Comment by gadmm** [comment]
~Real-world example: the failures in delimcc are solely due to the package defining CAML_NAME_SPACE, thereby not benefiting from the automatic inclusion of `compatibility.h`: http://check.ocamllabs.io/log/1575751838-4e17ba8dea3c1a1680cb8588d2395096ad6f867f/4.10/bad/delimcc.2018.03.16~

Here's something harder to address. The following shows what happens when you include some headers before defining `CAML_NAME_SPACE` and some headers after it: http://check.ocamllabs.io/log/1575751838-4e17ba8dea3c1a1680cb8588d2395096ad6f867f/4.10/bad/ocurl.0.9.0
It is tempting to blame the programmer, but it is going to be a very confusing issue, hard to debug. My present change to the documentation mildy addresses that by mentioning that the recommended way of defining `CAML_NAME_SPACE` is with an option passed to the preprocessor. (A better solution might be to change the includes in `caml/*.h` to make sure that domain_state.h and compatibility.h are always included together somehow.)

**Comment by gadmm** [comment]
~Sorry, I misread for delimcc. Delimcc defines `CAML_INTERNALS` and includes `memory.h`, which includes `compatibility.h` when `CAML_INTERNALS` is defined. This is in fact currently the only way one can benefit from `compatibility.h` when `CAML_NAME_SPACE` is defined. The only link with the current patch is that it aims to even out this irregularity.~ Thanks @kit-ty-kate for looking into this (#9202).

Edit: I misread again. See #9202. Sorry for the noise.

**Comment by gadmm** [comment]
This patch is indeed useful in making delimcc compile without a change (<https://github.com/ocaml/ocaml/pull/9202#issuecomment-568302282>).

**Comment by gadmm** [comment]
rebased

File: manual/manual/cmds/intf-c.etex [hunk 1]
```diff
@@ -182,26 +182,40 @@ section~\ref{s:c-custom}).}
 \entree{"caml/intext.h"}{operations for writing user-defined
 serialization and deserialization functions for custom blocks
 (see section~\ref{s:c-custom}).}
+\entree{"caml/signals.h"}{runtime lock, signals, and other
+  asynchronous callbacks (see
+  section~\ref{ss:c-process-pending-actions}).}
```
Comment (Octachron) [inline manual/manual/cmds/intf-c.etex:187]: @gadmm , would you mind If I pulled this documentation update in a separate PR?

Comment (gadmm) [reply manual/manual/cmds/intf-c.etex:187]: No problem, take all you need and add it to your PR

**Comment by gadmm** [comment]
The bug reported in this PR has been fixed differently at #9253, so I am closing this one. The second part of this PR proposed a clearer alternative to the CAML_API_VERSION proposals from other PRs. This part can be revived later if needed.
